### ISSUE TITLE
[Feature Req.] Warning when resolution is too low?

### ISSUE THREAD

**Issue by qyx268** [21cmfast/21cmFAST#76]
**Is your feature request related to a problem? Please describe.**
I was trying to do some debugging with large box. To speed it up, I was running very low resolution simulations. Then the program reports errors at this line

https://github.com/21cmfast/21cmFAST/blob/fa04f8f58ef7e289461f97836ee8471a69a83e4b/src/py21cmfast/src/SpinTemperatureBox.c#L1084

I checked. While `fcoll_int==(NSFR_low - 1)` is true, curr_dens << global_params.CRIT_DENS_TRANSITION. This is because `if( max_densities[R_ct]*zpp_growth[R_ct] > global_params.CRIT_DENS_TRANSITION )` is False, meaning the maximum density is even lower than 1.5 and caused by requesting very low resolution. 

**Describe the solution you'd like**

Maybe warn the user and terminate the program when very low-resolution has been requested? Otherwise, it might be confusing to them.

**Describe alternatives you've considered**

Of course, we can add features to circle around and allow very low resolution to be used...

**Additional context**

**Comment by steven-murray** [comment]
I think it's perfectly reasonable to warn/error in python if it is detected
that the resolution is too low. Can we exactly define (in python) when that
error will occur?

**Comment by BradGreig** [comment]
This is not trivial to catch on the python end. Its a function of box size and resolution. You would need to know what the maximum density in the box is before figuring out if you have sufficient resolution.

I recommend instead splitting up the if statement to be the following:
```
if(fcoll_int==(NSFR_low -1) && curr_dens <= ( global_params.CRIT_DENS_TRANSITION - 1e-4) {
    // Notice here, we take one step back in fcoll_int to account for the fact that we are near the end of the interpolation table.
    // We have entered here from a numerical rounding issue, thus evaluating this at one step back
    // should put us slightly beyond the end-point of the interpolation table. This however,
    // should be accurate enough as we are only beyond by the small rounding error.
    fcoll = log10_SFRD_z_low_table[R_ct][fcoll_int-1]*( 1 + (float)(fcoll_int-1) - dens_val ) + log10_SFRD_z_low_table[R_ct][(fcoll_int-1)+1]*( dens_val - (float)(fcoll_int-1) );
}
else if(fcoll_int==(NSFR_low -1) && fabs(curr_dens -global_params.CRIT_DENS_TRANSITION) < 1e-4) {
    // There can be instances where the numerical rounding causes it to go in here, rather than the curr_dens > global_params.CRIT_DENS_TRANSITION case
    // This checks for this, and calculates f_coll in this instance, rather than causing it to error
    dens_val = (curr_dens - fcoll_interp_high_min)*fcoll_interp_high_bin_width_inv;

    fcoll_int = (int)floorf( dens_val );

    fcoll = SFRD_z_high_table[R_ct][fcoll_int]*( 1. + (float)fcoll_int - dens_val ) + SFRD_z_high_table[R_ct][fcoll_int+1]*( dens_val - (float)fcoll_int );

}
else {
    fcoll_int_boundexceeded = 1;
}
```

This should enable it to deal with the issue without causing it to error.

**Comment by qyx268** [comment]
I agree with @BradGreig  that it's not easy to capture in python since you need to know the maximum density which changes with your initial conditions. I suggest taking the last value from the table directly rather than using the last two values to do interpolation and also adding a warning in this situation...

**Comment by steven-murray** [comment]
A couple of things:

1. The last clause (`fcoll_int_boundexceeded = 1`) -- does that end up raising an error?
2. Is this something required for v3? Is it urgent or can we push it down the line?

**Comment by qyx268** [comment]
> A couple of things:
> 
>     1. The last clause (`fcoll_int_boundexceeded = 1`) -- does that end up raising an error?
> 
never happened in my case. But I think we should do the same as for the low-density tables, e.g. using log10_SFRD_z_low_table when fcoll_int < 0...
>     2. Is this something required for v3? Is it urgent or can we push it down the line?
this seems to be a quick update if we do what @BradGreig suggested...

**Comment by BradGreig** [comment]
@steven-murray its there to capture an overstepping of the interpolation tables. It then returns an error message (through LOG_ERROR) and exits through a return(2). This, is one of the exit codes you put in place a while back.

@qyx268 I agree, that just taking the last value of the interpolation table should be sufficient. It'll be close enough to the actual value anyway.

As @qyx268 mentions, its a quick fix as its only a modification to the existing code. Thus, it should be able to go into v3 straightaway.

**Comment by steven-murray** [comment]
OK awesome, let's do it. 

I propose replacing the last `else` clause of @BradGreig 's example directly with 

```c
LOG_ERROR("I have overstepped my allocated memory for one of the interpolation tables for the fcoll/nion_splines");
return(2);
```

**Comment by steven-murray** [comment]
Actually, scrap that. If we can wait til the `better-exceptions` branch has been merged, we can actually _raise_ an error here.

**Comment by BradGreig** [comment]
@steven-murray I'd prefer it to remain as I had it to avoid it spitting out the same error multiple times (as its located within the loop). Capturing it as an int then spitting out an error at the end ensures its only seen once. This behaviour would also be preferable when the multithreading branch is merged as then you won't have instances of multiple threads trying to return similar errors.

**Comment by steven-murray** [comment]
Ah OK. Point taken. Actually, in the new exception handling system, the
error will be *raised* rather than just printed, and so the function will
instantly exit there and return to Python, which avoids the
multiple-printing issue. However, I'm not sure how this works under
multithreading...

**Comment by qyx268** [comment]
This can be closed after #37 and I just realized that I've added LOG_ERROR inside the loop for debugging during #37. But I guess we can wait for `better-exceptions` to solve that.
